kong/plugins/basic-auth/crypto.lua

```diff
@@ -8,7 +8,7 @@ local format = string.format
 -- Password is salted with the credential's consumer_id (long enough, unique)
 -- @param credential The basic auth credential table
 local function salt_password(credential)
-  return format("%s%s", credential.password, credential.consumer_id)
+  return format("%s%s", credential.password or "", credential.consumer_id)
```

Comment:
This makes sense as a safeguard. However, I was not able to trigger the issue you are talking about from a user's perspective (not in unit tests). i.e.:

Setting an empty password on a Consumer's credential:
```
$ http PATCH :8001/consumers/bob/basic-auth/8d2f6e55-c64c-4062-809b-25c70be70622 password=''
...
{
    "consumer_id": "689e311e-351b-4f17-8e4a-6a72547dfc4a",
    "created_at": 1520966590000,
    "id": "8d2f6e55-c64c-4062-809b-25c70be70622",
    "password": "7b3a51f9780906c381a389c93359518a6a4a6097",
    "username": "bob"
}
```

The `password` field is now made of `<empty string><consumer_id>`, like so:
```
$ echo -n "689e311e-351b-4f17-8e4a-6a72547dfc4a" | shasum 
7b3a51f9780906c381a389c93359518a6a4a6097  -
```

Now, the following request with an empty password successfully authenticates:
```
 $ http -v :8000/status/200 "Authorization:Basic `echo "bob:" | base64`"
```

Will you provide an example of the scenario you are describing (salting a credential in the request with `nil`)?

Another issue I observed though is when using JSON `null`, which this salting logic does not handle well either:

Setting the password to `null`:
```
$ http PATCH :8001/consumers/bob/basic-auth/8d2f6e55-c64c-4062-809b-25c70be70622 password:=null
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Date: Tue, 13 Mar 2018 19:18:34 GMT
Server: kong/0.12.3
Transfer-Encoding: chunked

{
    "consumer_id": "689e311e-351b-4f17-8e4a-6a72547dfc4a",
    "created_at": 1520966590000,
    "id": "8d2f6e55-c64c-4062-809b-25c70be70622",
    "password": "38aeadf84437ca9bdc46fc5a4be39f00581fd9ea",
    "username": "bob"
}
```

Which results in the following `password` field being hashed:
```
$ echo -n "userdata: NULL689e311e-351b-4f17-8e4a-6a72547dfc4a" | shasum
38aeadf84437ca9bdc46fc5a4be39f00581fd9ea  -
```

But requests coming from clients without a password will result in empty strings (as demonstrated above), hence will not verify against the hash stored in database. We should handle this case as well, since un-setting non-required fields via JSON `null` is officially supported by the Admin API.

This is using libpcre 8.41-5

Reply:
@thibaultcha This problem was showing itself to me because when I create the credential for the consumer, I'm only passing `username` in the query params. I don't set password to anything, so at creation time it comes in and gets encoded with NULL. 

The requests from clients without passwords are handling properly, but when they are compared against the digest with NULL in it, they're found not equal so the proxy doesn't allow access.

Reply:
Try to `curl -X POST /consumers/<consumer_id>/basic-auth -d '{"username": "foo"}'` and then make requests with credential.

Reply:
Ok, so I believe this is the case I commented about above (the second part of my comment about JSON `null`). In which case, this patch does nothing to prevent it, and the test case is wrong. `ngx.null`, aka `cjson.null` is not equal to `nil` in Lua, but is a lightuserdata.

Reply:
@thibaultcha I may just not be understanding where you're coming from. I just added  a log statement to the `encrypt` function in crypto.lua and also removed my nil check. Then I did the following steps multiple times and saw the logs below:

1. `curl -X POST localhost:8001/consumers/77e63858-f771-4929-9573-ba20d9074a48/basic-auth -H "Content-Type: application/json" -d '{"username": "myuser"}'`
2. Checked logs in `servroot/logs/error.log` and see:
```
2018/03/13 21:10:02 [debug] 6418#0: *1930 [lua] crypto.lua:20: encrypt(): nil77e63858-f771-4929-9573-ba20d9074a48

```

so `nil` is appended to the consumer_id to form the salt. 



crypto.lua
```
local resty_sha1 = require "resty.sha1"
local to_hex = require "resty.string".to_hex
local format = string.format

--- Salt the password
-- Password is salted with the credential's consumer_id (long enough, unique)
-- @param credential The basic auth credential table
local function salt_password(credential)
  return format("%s%s", credential.password, credential.consumer_id)
end

return {
  --- Encrypt the password field credential table
  -- @param credential The basic auth credential table
  -- @return hash of the salted credential's password
  encrypt = function(credential)
    local salted = salt_password(credential)
    ngx.log(ngx.DEBUG, salted)
    local digest = resty_sha1:new()
    assert(digest:update(salted))
    return to_hex(digest:final())
  end
}
```

Reply:
I get your point, you are making a `POST` and I was using `PATCH`. (in any case, this is still a good fallback to have).

From a `POST`, the schema validator will already handle a JSON `null` and will replace it with `nil`. But the following triggers the JSON null edge-case I was talking about:
```
$ curl -v -X PATCH \
  --urllocalhost:8001/consumers/689e311e-351b-4f17-8e4a-6a72547dfc4a/basic-auth/8d2f6e55-c64c-4062-809b-25c70be70622 \
  -H "Content-Type: application/json" \
  -d '{"password":null}'
```

Salted value:
```
salt_password(): userdata: NULL689e311e-351b-4f17-8e4a-6a72547dfc4a
```

Ultimately, this should be up to the schema_validator to better handle, but since it is on a deprecation path, it might be better to handle this here for now.

Maybe something like this:
```lua
local passwd = credential.password
if passwd == nil or passwd == ngx.null then
  passwd = ""
end
```

Reply:
Ah, I gotcha now. I'll update this to handle that case as well. Thanks for taking the time to explain :+1: